### PR TITLE
[mongo] Adds metrics for collections

### PR DESCRIPTION
**Pull request by gmmeyer**
### What does this PR do?

This PR adds some additional metrics from collections. It will now collect the following metrics from each collection (from the database that you're collecting from):

```
mongodb.collection.indexSizes
mongodb.collection.size
mongodb.collection.avgObjSize
mongodb.collection.count
mongodb.collection.capped
mongodb.collection.max
mongodb.collection.maxSize
mongodb.collection.storageSize
mongodb.collection.nindexes
```

### Motivation

Collection level metrics, like the amount of space that a collection is taking up, or the average object size, scoped to a specific collection, can be very important. If a certain collection is too large, maybe it should be split up. If the indexes on a collection are getting much, much larger than the others, maybe you want to split the collection. There are a lot of these things that can be important.

### Testing Guidelines

It tests to see whether it works in the most basic sense.

### CODE REVIEW

**Comment by gmmeyer** [comment]
The apache failure is not mine, and that's the only failure :)

File: checks.d/mongo.py [hunk 1]
```diff
@@ -355,6 +367,7 @@ class MongoDb(AgentCheck):
         'durability': DURABILITY_METRICS,
         'locks': LOCKS_METRICS,
         'wiredtiger': WIREDTIGER_METRICS,
+        'collection': COLLECTION_METRICS,
```
Comment (remh) [inline checks.d/mongo.py]: Does it need to be in the default metrics as well ? Looks like it's processed differently, no ?

File: checks.d/mongo.py [hunk 2]
```diff
@@ -875,3 +887,49 @@ def total_seconds(td):
...
+        # get collection level stats
+        try:
+            # Ensure that you're on the right db
+            db = cli[db_name]
+            # grab the collections from the database
+            coll_names = db.collection_names()
+            collection_metrics = []
+            for m in self.COLLECTION_METRICS:
+                # grab all the collections
+                collection_metrics.append(m.split('.')[1:])
```
Comment (remh) [inline checks.d/mongo.py]: Do we need to generate that list at each iteration ? It seems to be a constant to me.

Comment (gmmeyer) [inline checks.d/mongo.py]: Yup, I think you're right. I can generate it elsewhere.

File: checks.d/mongo.py [hunk 3]
```diff
@@ -875,3 +887,49 @@ def total_seconds(td):
...
+
+            # loop through the collections
+            for coll_name in coll_names:
+                # grab the stats from the collection
+                stats = db.command("collstats", coll_name)
+                # loop through the metrics
+                for m in collection_metrics:
```
Comment (remh) [inline checks.d/mongo.py]: Looking at the the content of COLLECTION_METRICS, collection_metrics will be a list of lists with a single element each.

File: checks.d/mongo.py [hunk 4]
```diff
@@ -875,3 +887,49 @@ def total_seconds(td):
...
+                    coll_tags = tags + ["db:%s" % db_name, "collection:%s" % coll_name]
+                    value = None
+                    try:
+                        # descend into the dict to grab the value we're looking for
+                        for c in m:
```
Comment (remh) [inline checks.d/mongo.py]: Looking at the the content of COLLECTION_METRICS, collection_metrics will be a list of lists with a single element each.

So there will be just one element in m

File: ci/mongo.rb [hunk 5]
```diff
@@ -70,6 +70,16 @@ def mongo_rootdir
...
+
+      bar_cmd = 'db.bar.save({1: []}); db.bar.save({})'
+      foo_cmd = 'db.foo.save({1: []}); db.foo.save({}); db.foo.save({1: []});'
+      sh %(#{mongo_rootdir}/bin/mongo\
+           --eval "#{bar_cmd}" 'localhost:37017/test'\
+           >> $VOLATILE_DIR/mongo.log)
+      sh %(#{mongo_rootdir}/bin/mongo\
+          --eval "#{foo_cmd}" 'localhost:37017/test'\
+          >> $VOLATILE_DIR/mongo.log)
+      sleep_for 30
```
Comment (remh) [inline ci/mongo.rb]: why do we need a 30s sleep ?

File: checks.d/mongo.py (hunk 4 above)
Comment (gmmeyer) [inline checks.d/mongo.py]: Yea, with the ones I added, that's correct, but there's others that we could grab from the dict that would have more.

**Comment by remh** [comment]
We should not do it for every collection by default. Can you do something similar to the postgres check where we list the tables we want more metrics on (here it would be the collections not the tables)

**Comment by degemer** [comment]
👍

**Pull request merged**